server/datastore/mysql/software.go

```diff
@@ -719,19 +708,53 @@ func (ds *Datastore) AllSoftwareWithoutCPEIterator(ctx context.Context, excluded
 	return &softwareIterator{rows: rows}, nil
 }
 
-func (ds *Datastore) AddCPEForSoftware(ctx context.Context, software fleet.Software, cpe string) error {
-	_, err := addCPEForSoftwareDB(ctx, ds.writer, software, cpe)
-	return err
+func (ds *Datastore) UpsertSoftwareCPEs(ctx context.Context, cpes []fleet.SoftwareCPE) (int64, error) {
+	var args []interface{}
+
+	if len(cpes) == 0 {
+		return 0, nil
+	}
+
+	values := strings.TrimSuffix(strings.Repeat("(?,?),", len(cpes)), ",")
+	sql := fmt.Sprintf(
+		`INSERT INTO software_cpe (software_id, cpe) VALUES %s ON DUPLICATE KEY UPDATE cpe = VALUES(cpe)`,
+		values,
+	)
+
+	for _, cpe := range cpes {
+		args = append(args, cpe.SoftwareID, cpe.CPE)
+	}
+	res, err := ds.writer.ExecContext(ctx, sql, args...)
+	if err != nil {
+		return 0, ctxerr.Wrap(ctx, err, "insert software cpes")
+	}
+	count, _ := res.RowsAffected()
+
+	return count, nil
 }
 
-func addCPEForSoftwareDB(ctx context.Context, exec sqlx.ExecerContext, software fleet.Software, cpe string) (uint, error) {
-	sql := `INSERT INTO software_cpe (software_id, cpe) VALUES (?, ?)`
-	res, err := exec.ExecContext(ctx, sql, software.ID, cpe)
+func (ds *Datastore) DeleteSoftwareCPEs(ctx context.Context, cpes []fleet.SoftwareCPE) (int64, error) {
+	if len(cpes) == 0 {
+		return 0, nil
+	}
+
+	sql := fmt.Sprintf(
+		`DELETE FROM software_cpe WHERE (software_id) IN (%s)`,
```

Comment:
Any reason not to use `sqlx.In` like in the other methods?

Reply:
Done in https://github.com/fleetdm/fleet/pull/10514/commits/8c2e90168b77aaf402e96e4b18c5e5bc2ce018c8